architecture-decision-records/model-serving/ODH-ADR-MS-0004-ai-gateway-tenancy-discovery.md

```
@@ -0,0 +1,140 @@
+TBD
+
+## Alternatives
+
+##### Tenants cache stored in Postgres
+  The problem with this approach is that while it greatly reduces the number of hits to the kube-api server, it creates more complexity for the service and potentially adds a risk since there is no longer a single source of truth for the tenancy information, but rather two: kube-api and Postgres, and these need to be kept in sync.
+
+##### maas-controller continuously informs the discovery service of any change in the tenants/gateway state
+   The problem with this is the increase in complexity. Also, it is not enough for the maas-controller to send an update request to the discovery service because the service may have N replicas at one point in time, and all replicas would need to be informed about the changes. This can be mitigated with the above alternative of using a DB layer, but again, this would increase complexity. 
+
+##### Signal cache refresh via a storage (Postgress, Redis, etc)
+    In this scenario maas-controller can singnal that "something changed' in the AITenant or Gateway byt making an internal request to the discovery-service. This request may land in any POD replica and this will record that the state chageds in a persistent store. Then, each replica can watch this flag and when detected each replica refreshes its cache. This can be more efficient that time based cache refresh as changes in tenants may happen rarely and the discovery service does not really need to hit the kube-api server periodically. However this is more complex to implement than time base cach refresh and may be considered in a later phase.
```

markdownlint-cli2 (0.22.1)

[warning] Code block style
Expected: fenced; Actual: indented

(MD046, code-block-style)